### ISSUE TITLE
Tracking issue: 32bit x86 targets have non-compliant floating point behavior

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust#114479]
There are two ways in which our 32bit x86 targets deviate from the desired floating point behavior:
- On all of them, returning a floating-point number by-value can change the bits of their NaN payloads. This is caused by using x87 registers to pass return values, as mandated by the ABI for this target.
- On some of them, floating-point operation can return results that are rounded in different ways than they should, and results can be "inconsistent": depending on whether const-propagation happened, the same computation can produce different results, leading to a program that seemingly contradicts itself. This is caused by using x87 instructions to perform floating-point arithmetic, which do not accurately implement IEEE floating-point semantics.

These are known and long-standing problems, and very hard to fix. The purpose of this issue mostly is to document their existence and to have a common place to reference for these problems.

Some ideas that have been floated for fixing these problems:
- The first problem can never be fully fixed, since the use of floating-point registers for return values is mandated by the platform ABI. However, it *can* be fixed for Rust's own ABI, which can just choose to use different registers. That would make the problem much less likely to appear in practice.
- The second problem can be properly fixed by emitting different instruction sequences for floating-point operations that emulate the expected rounding behavior using x87 instructions. This will likely require changes deep in LLVM's x86 backend. This is [what Java does](https://open-std.org/JTC1/SC22/JSG/docs/m3/docs/jsgn326.pdf).

**Comment by DemiMarie** [comment]
Can we just drop support for x86 without SSE2 or fall back to software floating point?

**Comment by RalfJung** [comment]
> Can we just drop support for x86 without SSE2 

We currently have the following targets in that category:
- i586-pc-windows-msvc
- i586-unknown-linux-gnu
- i586-unknown-linux-musl

They are all tier 2. I assume people added them for a reason so I doubt they will be happy about having them dropped. Not sure to what extent floating point support is needed on those targets, but I don't think we have a concept of "target without FP support".

>  or fall back to software floating point?

In theory of course we can, not sure if that would be any easier than following the Java approach.

**Comment by programmerjake** [comment]
> > or fall back to software floating point?
> 
> In theory of course we can, not sure if that would be any easier than following the Java approach.

it should be much easier since LLVM already supports that, unlike the Java scheme (afaik)

https://gcc.godbolt.org/z/cahK4c8z7

**Comment by RalfJung** [comment]
I wonder if there's something that could be done about the fact that this also affects tier 1 targets with custom (stable) flags such as `-C target-cpu=pentium`. Do we still want to consider that a tier 1 target in itself? Is there a way that we can just reject flags that would disable SSE2 support, and tell people to use a different target instead?

**Comment by DemiMarie** [comment]
@RalfJung What about forcing non-SSE2 targets to software floating point?  That must be supported anyway because of kernel code.

**Comment by RalfJung** [comment]
Yeah that's an option listed above, since you already proposed it before. I have no idea how feasible it is. People are very concerned about the softfloat support for f16/f128 leading to code bloat and whatnot, so the same concerns would likely also apply here.

AFAIK the kernel code just doesn't use floats, I don't think they have softfloats?

**Comment by briansmith** [comment]
Just to clarify, this is really only for 32-bit x86 non-SSE targets, and doesn't affect x86-64 non-SSE2 targets like x86-64-unknown-none?

> Can we just drop support for x86 without SSE2

I would guess we'll need to support i686-unknown-none like x86-64-unknown-none for use in operating system kernels like Linux that don't allow vector registers to be used (without extra work) even when the hardware has them.

> or fall back to software floating point?

That appears to be what `x86_64-unknown-none` does, one of the non-SSE2-by-default x86-64 targets. At least the table at https://doc.rust-lang.org/beta/rustc/platform-support.html says "softfloat", though the detailed target documentation at https://doc.rust-lang.org/beta/rustc/platform-support/x86_64-unknown-none.html doesn't mention softfloat.

**Comment by RalfJung** [comment]
> Just to clarify, this is really only for 32-bit x86 non-SSE targets, and doesn't affect x86-64 non-SSE2 targets like x86-64-unknown-none?

x86-64-unknown-none uses softfloats so it should not be affected.
I don't think there are any x86-64 hardfloat targets without SSE.

> I would guess we'll need to support i686-unknown-none like x86-64-unknown-none for use in operating system kernels like Linux that don't allow vector registers to be used (without extra work) even when the hardware has them.

Right, I should have been more precise -- I meant for hardfloat targets.

**Comment by Jules-Bertholet** [comment]
> We could set the FPU control register to 64bit precision for Rust programs, and require other code to set the register in that way before calling into a Rust library.

This is not sufficient to ensure full IEEE754 compliance,

**Comment by RalfJung** [comment]
What's described there sounds different. That is using `volatile` to force rounding from "extended precision" to "double precision" after each operation. But of course the actual operation is still done with extended precision, which leads to double-rounding, which leads to wrong results.

The proposal you quoted was to switch x87 precision such that the operation is performed with double-precision to begin with, entirely avoiding double rounding.

It is possible that the x87 has further issues that make this not work, but the link you posted does not seem to even mention idea if changing the FPU control register to get a different precision, so as far as I can see it doesn't provide any evidence that setting the x87 to 64bit precision would lead to incorrect results.

**Comment by Jules-Bertholet** [comment]
From the post (emphasis mine):

> Floating-point calculations done with x87 FPU instructions are done in extended-precision registers, **even when the processor is in double-precision mode** and the variables being operated on are double-precision.

**Comment by RalfJung** [comment]
Hm, my understanding was that switching the FPU mode to 64bit would solve the problem. But of course it's possible that x87 screws up even when explicitly asked for IEEE 754 64bit arithmetic. :shrug: 

Requiring the FPU to be in 64bit mode is anyway not a realistic option, I listed it just for completeness' sake.

**Comment by RalfJung** [comment]
Does the Java encoding behave correctly for denormals?

That it can express too many exponents wouldn't be a problem if the extra precision doesn't lead to different rounding results.

**Comment by Jules-Bertholet** [comment]
Yes, Java behaves correctly in all cases. It scales the exponent of one of the arguments before multiplication and division to ensure that where a 64-bit op would evaluate to a denormal, the 80-bit op does the same (and then it scales the exponent of the result back to what it should be). This is all described in the PDF linked from OP.

**Comment by RalfJung** [comment]
Wow, that's a great example. It even works in entirely safe code. Impressive.

We probably have to upgrade this issue to I-unsound then. Is there an upstream LLVM issue for this miscompilation?

**Comment by Muon** [comment]
That *is* impressive. It looks closely related to https://github.com/llvm/llvm-project/issues/44218. Perhaps it's yet another example to toss on that pile?

**Comment by taiki-e** [comment]
> As loading and storing `f32`s and `f64`s to and from the x87 floating point stack quiets signalling NaNs

(I have not tested it, but I wonder if this x87's behavior actually affects not only floats, but also the i586's 64-bit atomic load/store, which [uses x87 load/store instructions](https://github.com/llvm/llvm-project/blob/llvmorg-18.1.2/llvm/test/CodeGen/X86/atomic-load-store-wide.ll#L61-L62).)

**Comment by apiraino** [comment]
Nominating for t-compiler discussion.

This tracking issue shows that we have targets that intersect tier platforms support in different ways. For example i686 are tier 1 but "non-SSE2" are tier 2. All these differences are not apparent [in our documentation](https://doc.rust-lang.org/rustc/platform-support.html). 

So as discussed on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23114479.20Tracking.20issue.3A.2032bit.20x86.20targets.20without.20SSE2.20have.E2.80.A6/near/434964647)  there are probably a number of questions:
- targets that just don't have SSE2, like i586 -- these are tier 2, maybe critical codegen bugs are "fine" there?
- tier 1 targets with SSE2 disabled. we dont have a way to say "that's actually a tier 2 situation" so maybe we should just emit a hard error? or at least a warning?
- (to which I also add): if these issues stem from LLVM, can we faithfully represent a realistic situation or should we just point at LLVM and say "we just do what they do"?

(please feel free to add context/correct how I represent the problem, thanks!)

@rustbot label I-compiler-nominated

**Comment by RalfJung** [comment]
In a sense the issue stems from LLVM, yeah -- x86 without SSE seems to be a poorly supported target. Even in the best of cases, f64 just behaves incorrectly on that target (and f32, to a lesser extent), and then it turns out things are actually a lot worse.

Altogether I think there are enough issues with floating point on x86 without SSE (this one, and also https://github.com/rust-lang/rust/issues/115567) that IMO we should say that tier 1 hardfloat targets require SSE period. It is already the case that using feature flags to turn a hardfloat target into a softfloat target is unsound (Cc https://github.com/rust-lang/rust/issues/116344), and we should simply hard-error in those cases (e.g. disabling the x87 feature on any hardfloat x86 target). IMO we should do the same when disabling SSE/SSE2 on an i686 target.

**Comment by apiraino** [comment]
@rustbot label -I-compiler-nominated

Discussed in compiler triage (on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-04-25/near/435399386)). Copying here the [summary](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-04-25/near/435408565):
- we acknowledge this is a problem
- we should update target docs to contain a mention of this issue (non-SSE2 x86 codegen is basically unsound)
- a MIR lint to detect float ops when using these broken targets would help users
- a lint to detect mismatched -Ctarget-features between dependencies would help with -Ctarget-feature=-sse and similar cases for other platforms
- **Current Tier 1 x86 targets require SSE-based floats at minimum** (i.e. not softfloats)

**Comment by keith-packard** [comment]
Does Rust expose the difference between representation and evaluation method? In C, FLT_EVAL_METHOD` is used by applications to detect platforms on which the computations are performed at a higher precision than the underlying value representations. 

This doesn't completely solve the problem as 8087 ABI also does representation conversion across function call boundaries so passing (or returning) a sNaN will have that transformed into qNaN and an exception.

It's a stretch to claim that the 8087 FPU has 'unsound' floating point behavior; it's all quite compliant with the IEEE 754 specification, if you squint hard enough. Changing the ABI to pass float/double without a representation change would resolve the worst of the problems I've found, and that's not a hardware issue.

**Comment by keith-packard** [comment]
Yeah, it sounds like rust cannot support other values for `FLT_EVAL_METHOD`, and so doesn't support these kinds of corner cases in the IEEE 754 spec. I suspect Rust and the x87 FPU are just not going to get along. Even if you kludge around the sNaN->qNaN adventures, you're still going to suffer from double rounding as every computation is first rounded to 80 bits and then to 32 or 64. 

Having Rust produce results that depend on the underlying hardware seems antithetical to some pretty fundamental language goals. But, that would mean abandoning the x87 FPU entirely, and that doesn't seem feasible even though sse2, which provides the necessary native 32/64 binary format support is nearly old enough to serve in the US house of representatives.

I'd encourage someone to figure out how to tell applications that the underlying hardware doesn't work the way they expect. Mirroring some version of FLT_EVAL_METHOD from C would at least be easy and sufficient for this hardware.

**Comment by RalfJung** [comment]
> I'd encourage someone to figure out how to tell applications that the underlying hardware doesn't work the way they expect. Mirroring some version of FLT_EVAL_METHOD from C would at least be easy and sufficient for this hardware.

As @beetrees explained, it's not just that the underlying hardware works differently and that bleeds into language semantics. It's that Rust's primary backend, LLVM, *assumes* the underlying hardware to work the standard way -- the examples @beetrees referenced demonstrate that there is no reliable way to program against this hardware in any compiler that uses LLVM as its backend.

To my knowledge, nobody on the LLVM side really cares about this. So until that changes it is unlikely that we'll be able to improve things in Rust here.

> Even if you kludge around the sNaN->qNaN adventures, you're still going to suffer from double rounding as every computation is first rounded to 80 bits and then to 32 or 64.

It's possible to get around that using the approach Java used, IIUC. It does cost some performance but it avoids the semantic inconsistencies.

**Comment by DemiMarie** [comment]
There are legitimate use-cases for running with the FPU in flush-to-zero + denormals-are-zero mode, even on mainstream targets like x86-64.  Processing denormals outside of FTZ+DAZ mode requires a microcode assist that causes ~100 cycles IIRC.  In this application, denormal numbers correspond to sounds that are (IIUC) below the noise floor, so flushing them to zero is harmless.  Failing to meet deadlines because of the microcode assist is a bug.

That’s separate but related to this issue.

**Comment by Muon** [comment]
> For f32, double-rounding does not affect results.
This is only true for some operations, and then only if a single step is performed. It is not true for addition/subtraction, and it is not true even for other operations if multiple operations are performed at 80-bit precision. 

> There are legitimate use-cases for running with the FPU in flush-to-zero + denormals-are-zero mode, even on mainstream targets like x86-64.
Yes, but LLVM's support for anything except standard IEEE floating point arithmetic is broken or absent, so Rust can't hope to support this until LLVM does.

**Comment by RalfJung** [comment]
> There are legitimate use-cases for running with the FPU in flush-to-zero + denormals-are-zero mode, even on mainstream targets like x86-64.

Flushing denormals isn't even permitted by the IEEE standard. So while I am not doubting that that is something people may want to do, supporting it requires new language features (and, depending on the constraints, work on the LLVM side). Please take that to another thread.

> and it is not true even for other operations if multiple operations are performed at 80-bit precision.

Indeed the rounding down to `f32` has to happen after each operation.

**Comment by apiraino** [comment]
WG-prioritization assigning priority (see [Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23129880.20ARM.20NEON.20intrinsics.20are.20unsound.20due.20to.20subnormal.20fl.E2.80.A6) for the related issue #129880).

@rustbot label -I-prioritize +P-medium

**Comment by RalfJung** [comment]
Given that we have no clear plan for how this issue could be fixed, I wonder if it would make sense to demote the affected targets to tier 3, and instead cover the "no SSE" usecase with tier 2 targets that use softfloats.

**Comment by apiraino** [comment]
> Given that we have no clear plan for how this issue could be fixed, I wonder if it would make sense to demote the affected targets (i586-unknown-linux-gnu, i586-unknown-linux-musl) to tier 3, and instead cover the "no SSE" usecase with tier 2 targets that use softfloats.

Linking again [the notes](https://github.com/rust-lang/rust/issues/114479#issuecomment-2082686372) from the T-compiler discussion of last year. What's the current status and were any of the above points implemented? (didn't yet check, might try to do if nobody jumps in before me).

**Comment by RalfJung** [comment]
> we should update target docs to contain a mention of this issue (non-SSE2 x86 codegen is basically unsound)

This has been done a while ago:

![Image](https://github.com/user-attachments/assets/b0e56703-5de2-4960-bc31-1da04b297512)

> a MIR lint to detect float ops when using these broken targets would help users

I'm not aware of any movement in this direction. We'd have to also cover many of the float methods in std, so this lint would trigger a lot. Likely users would miss it since it occurs in dependencies.

> a lint to detect mismatched -Ctarget-features between dependencies would help with -Ctarget-feature=-sse and similar cases for other platforms

I don't quite understand this one; mismatches of the `sse` target feature are completely fine per se. There are some ABI issues but those are entirely orthogonal to this one.

> Current Tier 1 x86 targets require SSE-based floats at minimum

This has been implemented recently.

**Comment by RalfJung** [comment]
@Fabian-Gruenbichler [shared](https://github.com/rust-lang/rust/issues/136495#issuecomment-2771464137) the good news that Debian will not be affected by this any more since they are bumping their x86-32 baseline to match ours. :)

**Comment by RalfJung** [comment]
The best place would have been https://github.com/rust-lang/rust/issues/128288 6 months ago. ;) Right now, I'd say open a new issue and mention it in that tracking issue.

**Comment by LekKit** [comment]
I have stumbled upon a similar issue in my project, albeit not specifically related to Rust or LLVM, and have (somewhat) solved it.

My emulator uses native host FPU to emulate conforming RISC-V floating point unit, yet it was subject to same problems with 8087 FPU discussed here (Signaling NaN / NaN payloads not preserved between 8087 loads/stores, excess intermediate precision). There were other issues which i also solved, but they are out of scope here.

I solved it by not even using native FP types around the codebase, replacing them with 32/64 bit integer typedefs which carry the IEEE754 representation. They are freely passed around as integers on compiler IR / hardware level.

Only around actual floating-point computations I bitcast the integer-typedef into a native floating-point type. That is fine, as IEEE754 doesn't guarantee preserving NaN payload across an operation. The compilers (both GCC and Clang) are smart enough to recognize where to put an `fld/fst` or integer `mov` in almost every case, optimizing away excess bitcasts, plus I get correctly rounded result (to f32 or f64) after each operation.

Sure, this is somewhat slower than "naive" 8087 usage, but what matters is it was fully correct for my usecase, preserves passing around arbitrary NaN payloads, and is in fact orders of magnitude faster than softfloat.
So it might be worth considering this idea either on LLVM side, or generate "floats as integer types" from Rust to LLVM IR on i686, bitcasting whenever needed around actual computations.

One issue this does not solve, is returning arbitrary NaN payloads via 8087 ABI return. But maybe rust-specific ABI change for i686 could fix that, leaving only Rust<->C calls affected.

**Comment by RalfJung** [comment]
We still have to use float types for `extern "C"` function signatures to ensure the right ABI is used. Also, LLVM will probably optimize away bitcasts when writing a series of FP operations. In an interpreter, LLVM can't know what the actual program is so there's not much to optimize; this changes quite a bit when writing more concrete code. So, while this helps for some situations, it's far from a full fix.

**Comment by johnwcowan** [comment]
GCC has had this bug reported hundreds of times, so it's not an LLVM-specific problem.  The theoretical resolution is that you can have either fast or correct, but not both.  In practice, the resolution is to say that non-SSE x86 CPUs aren't supported any more: they don't work right and they never will, and anyway almost nobody has one.

**Comment by ssokolow** [comment]
> and anyway almost nobody has one

Bit of a shame to look at it like that. I've been meaning to explore the viability of using [Rust9x (UNOFFICIAL "Tier 4" Rust target for Windows 9x/Me/NT/2000/XP/Vista)](https://github.com/rust9x/rust/wiki) on my various Windows 98SE and XP (and possibly later dual-booting to 95/98/ME/NT4/2000) retro-hobby machines with CPUs like a 133MHz Pentium, 400MHz VIA Eden ESP 4000, 800 MHz VIA C7, 1GHz VIA C7, etc.

(And I'm part of the [Investigate Win32s support](https://github.com/rust9x/rust/discussions/48) discussion on it, because it'd be neat and potentially useful to be able to write Windows 3.11 for Workgroups stuff in Rust.)

...and who knows how many pre-SSE machines I'll have once I start to practice recapping motherboards using a bunch of capacitor-plagued machines I have stacked up under the basement stairs. They're not *all* Pentium 4s.

**EDIT:** RalfJung, Understood... but I wish you'd said that in the 8 hour gap *before* I got to be the one with egg on my face.

**Comment by RalfJung** [comment]
Please stop the off-topic discussion about how many people care about these systems. The fact is, for many years nobody was motivated and able to fix this in either LLVM or GCC. It is not worth discussing here why that may be the case. Only reply if you have new information about the actual issue, or want to work on it.

**Comment by DemiMarie** [comment]
What about simply refusing to target such machines?  In other words, always require SSE2 or softfloat, and in the latter case refuse to call any foreign function with a floating point argument not behind a pointer (because the ABI used would be wrong).

Or just unconditionally fail the build.

**Comment by bjorn3** [comment]
On 32bit x86 the ABI requires storing floats in x87 registers, so the only way to fully fix this issue is getting rid of 32bit x86 entirely, which is not an option at all for the forseeable future. We already avoid x87 registers in the Rust ABI irrespective of sse2, but can't fix the C ABI.

**Comment by RalfJung** [comment]
> the only way to fully fix this issue is getting rid of 32bit x86 entirely

That's not true; as discussed above, one could generate code that stores/loads an f32/f64 in an x87 register while preserving the NaN payload. Someone "just" has to implement this in LLVM.
